### PR TITLE
Use cachable flash address space while reading on infineon xmc4xxx

### PR DESCRIPTION
**Pull request by talih0**
The infineon xmc4xxx series has two ways to access flash: One is the
cachable address space at 0x8000000 which can use pre-fetched data to
reduce flash access latency; The second is uncached space at 0xc000000
which is mainly used for write and erase operations.
The two address spaces point to the physical flash.
    
This patch adds support for using the cachable read access by setting
it as the virtual address in the linker script for the rom section.

### CODE REVIEW

**Comment by talih0** [comment]
I've split up the pull request to simplify review and fixed some compliance checks.

File: dts/arm/infineon/xmc4500.dtsi [hunk 1]
```diff
@@ -21,5 +21,5 @@
 };
 
 &flash0 {
-	reg = <0xc000000 DT_SIZE_M(1)>;
+	reg = <0xc000000 DT_SIZE_M(1) 0x8000000 DT_SIZE_M(1)>;
```
Comment (parthitce) [inline dts/arm/infineon/xmc4500.dtsi:24]: So the idea here is to use cached address space always? If so, I need to read back the TRM about flushing the cache during abnormal scenarios like reset.

File: include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld [hunk 2]
```diff
@@ -207,8 +208,9 @@ SECTIONS
 #include <zephyr/linker/cplusplus-rom.ld>
 
 	__rodata_region_end = .;
-	MPU_ALIGN(__rodata_region_end -__rom_region_start);
+	MPU_ALIGN(__rodata_region_end - ADDR(rom_start));
```
Comment (parthitce) [inline include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld:211]: IMO these changes split into another PR to keep the arch changes independent of xmc?

File: dts/arm/infineon/xmc4500.dtsi (hunk 1 above)
Comment (talih0) [reply dts/arm/infineon/xmc4500.dtsi:24]: > 

That's how it's done in the Infineon linker script. I didn't see in the manual (or example code startup code) about handling these cases. 

There can be some cache coherency problems when writing/erasing cache and then reading from cached address space..  (but we don't have a flash driver yet):

Screenshot_2022-07-03_10-12-48

Comment (talih0) [reply dts/arm/infineon/xmc4500.dtsi:24]: > So the idea here is to use cached address space always?

The user can also go back to the previous version (using uncached address space) by setting:
reg = <0xc000000 DT_SIZE_M(1) 0xc000000 DT_SIZE_M(1)>;

File: include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld (hunk 2 above)
Comment (talih0) [reply include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld:211]: Yes, I think that's a good idea. For now, I'd like to keep the XMC4xxx changes here to add some context to the linker changes. I'll split up the PR in a few days.

Comment (parthitce) [reply include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld:211]: @talih0 possible to split this up?

Comment (talih0) [reply include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld:211]: > @talih0 possible to split this up?

I've put these changes in a separate PR.

**Comment by talih0** [comment]
I found some more linker snippets that need an ALIGN(4) to sync up the vma and lma addresses. I'll submit a separate PR for those changes. 
I'll set this PR to draft until those changes are merged.
